src/common/src/types/datetime.rs

```diff
@@ -495,18 +491,24 @@ impl Timestamp {
     }
 
     pub fn from_protobuf(cur: &mut Cursor<&[u8]>) -> ArrayResult<Timestamp> {
-        let micros = cur
+        let secs = cur
             .read_i64::<BigEndian>()
-            .context("failed to read i64 from Timestamp buffer")?;
+            .context("failed to read i64 from Time buffer")?;
+        let nsecs = cur
+            .read_u32::<BigEndian>()
+            .context("failed to read u32 from Time buffer")?;
 
-        Ok(Timestamp::with_micros(micros)?)
+        Ok(Timestamp::with_secs_nsecs(secs, nsecs)?)
     }
 
-    /// Although `Timestamp` takes 12 bytes, we drop 4 bytes in protobuf encoding.
     pub fn to_protobuf<T: Write>(self, output: &mut T) -> ArrayResult<usize> {
-        output
-            .write(&(self.0.and_utc().timestamp_micros()).to_be_bytes())
-            .map_err(Into::into)
+        let timestamp_size = output
+            .write(&(self.0.and_utc().timestamp()).to_be_bytes())
+            .map_err(Into::<ArrayError>::into)?;
+        let timestamp_subsec_nanos_size = output
+            .write(&(self.0.and_utc().timestamp_subsec_nanos()).to_be_bytes())
+            .map_err(Into::<ArrayError>::into)?;
+        Ok(timestamp_subsec_nanos_size + timestamp_size)
```

Comment:
This will break the compatibility of Timestamp stored in protobuf. Not sure whether it's used somewhere in our catalog.

Reply:
One of the ways to keep **backward** compatibility is: using a special value such as `-1` as a placeholder indicating the new format. That is

- If read `-1`, parse the subsequent 12 bytes in new format
- If read otherwise, consider it as a timestamp_micros (old format)